imap_processing/lo/l0/data_classes/science_direct_events.py

```diff
@@ -0,0 +1,213 @@
+        """Initialize Science Direct Events Data class."""
+        set_attributes(self, packet)
+        self.software_version = software_version
+        self.packet_file_name = packet_file_name
+        self.ccsds_header = CcsdsData(packet.header)
+        # TODO: Is there a better way to initialize these arrays?
+        self.TIME = np.array([])
+        self.ENERGY = np.array([])
+        self.MODE = np.array([])
+        self.TOF0 = np.array([])
+        self.TOF1 = np.array([])
+        self.TOF2 = np.array([])
+        self.TOF3 = np.array([])
+        self.CKSM = np.array([])
+        self.POS = np.array([])
+        self._decompress_data()
+
+    def _decompress_data(self):
+        """Decompress the Lo Science Direct Events data."""
+        data = BinaryString(self.DATA)
+        for _ in range(self.COUNT):
+            self._parse_data(data)
+
+    def _parse_data(self, data: BinaryString):
+        """Parse the TOF data.
+
+        TOF data is parsed and the direct event data class
+        attributes are set.
+
+        Parameters
+        ----------
+        data : BinaryString
+            BinaryString object containing the TOF binary.
+        """
+        # The first 4 bits of the binary data are used to
+        # determine which case number we are working with.
+        # The case number is used to determine how to
+        # decompress the TOF values.
+        case_number = int(data.next_bits(4), 2)
+
+        # time, energy, and mode are always transmitted.
+        time = int(data.next_bits(DATA_BITS.TIME), 2)
+        self.TIME = np.append(self.TIME, time)
+
+        energy = int(data.next_bits(DATA_BITS.ENERGY), 2)
+        self.ENERGY = np.append(self.ENERGY, energy)
+
+        mode = int(data.next_bits(DATA_BITS.MODE), 2)
+        self.MODE = np.append(self.MODE, mode)
+
+        # Case decoder indicates which parts of the data
+        # are transmitted for each case.
+        case_decoder = CASE_DECODER[(case_number, mode)]
+
+        # TOF values are not transmitted for certain
+        # cases, so these can be initialized to the
+        # CDF fill val and stored with this value for
+        # those cases.
+        tof0 = GlobalConstants.DOUBLE_FILLVAL
+        tof1 = GlobalConstants.DOUBLE_FILLVAL
+        tof2 = GlobalConstants.DOUBLE_FILLVAL
+        tof3 = GlobalConstants.DOUBLE_FILLVAL
+        cksm = GlobalConstants.DOUBLE_FILLVAL
+        pos = GlobalConstants.DOUBLE_FILLVAL
+
+        # Check the case decoder to see if the TOF field was
+        # transmitted for this case. Then grab the bits from
+        # the binary and perform a bit shift to the left. The
+        # data was packed using a right bit shift (1 bit), so
+        # needs to be bit shifted to the left (1 bit) during
+        # unpacking.
+        if case_decoder.TOF0:
+            tof0 = int(data.next_bits(DATA_BITS.TOF0), 2) << DE_BIT_SHIFT
+        self.TOF0 = np.append(self.TOF0, tof0)
```

Comment:
If you take the previous suggestion of creating the array initially with fill values, where index would need to be passed in for your current case.
```suggestion
        if case_decoder.TOF0:
            self.TOF0[index] = int(data.next_bits(DATA_BITS.TOF0), 2) << DE_BIT_SHIFT
```